File: src/core/breeding.ts
/**
 * Cruza: dos genomas dan uno nuevo.
 *
 * ---
 *
 * LA DECISIÓN CENTRAL: se cruza por GEN, no por bit.
 *
 * Lo intuitivo sería mezclar bit a bit, y está mal. El genoma es un número
 * empaquetado: `hue` son los bits 24 a 31. Si de esos ocho se toman cuatro de
 * cada padre, el resultado no es "un color intermedio" — es un color al azar
 * que no se parece a ninguno de los dos. Lo mismo con cada gen: mezclar bits
 * dentro de un campo destruye su significado.
 *
 * Cruzando por gen, cada campo viene entero de uno de los padres. El hijo tiene
 * los ojos de uno y el color del otro, y eso se VE. Es la diferencia entre que
 * la herencia sea un mecanismo o sea ruido.
 *
 * La mutación sí actúa bit a bit, y ahí es lo correcto: un bit dado vuelta en
 * `hue` corre el tono un poco, uno en `lineage` cambia el linaje entero. Eso es
 * exactamente lo que tiene que hacer una mutación.
 *
 * ---
 *
 * Módulo puro y determinista: los mismos padres con el mismo `nonce` dan
 * siempre el mismo hijo. El azar entra por el `nonce`, que lo elige quien
 * llama, así que la función se puede testear sin trucos.
 */

import { GENOME_LAYOUT } from "./genome.ts";
import { deriveSeed, mulberry32 } from "./rng.ts";
import type { CreatureState } from "./simulation.ts";

/** Cuántos bits se espera que muten, en promedio, sobre los 64. */
const MUTACIONES_ESPERADAS = 2.5;

/** Ocho horas de descanso entre cruzas. */
export const CRUZA_COOLDOWN_MS = 8 * 60 * 60 * 1000;

/**
 * Vínculo mínimo para poder cruzar.
 *
 * Con el tope diario de 12, llegar a 20 lleva por lo menos dos días de
 * atención. Es deliberado: ata la cruza al bucle de cuidado en vez de
 * convertirla en un atajo para saltárselo.
 */
export const CRUZA_MIN_VINCULO = 20;
export const CRUZA_MIN_SALUD = 50;

export type Origen = "A" | "B" | "mutado";

export interface Cruza {
  seed: bigint;
  /** De dónde salió cada gen. Permite mostrar a quién salió el hijo. */
  herencia: Record<string, Origen>;
  /** Cuántos bits mutaron. */
  mutaciones: number;
}

/**
 * Combina dos genomas.
 *
 * Es simétrica: cruzar A con B da lo mismo que cruzar B con A. El par
 * determina el hijo, no el orden en que los elegiste.
 */
export function cruzar(seedA: bigint, seedB: bigint, nonce: number): Cruza {
  // Se ordenan para sembrar el azar, que es lo que la vuelve simétrica.
  const [menor, mayor] = seedA <= seedB ? [seedA, seedB] : [seedB, seedA];
  const rng = mulberry32(deriveSeed(menor, `cruza:${mayor}:${nonce}`));

  const herencia: Record<string, Origen> = {};
  let hijo = 0n;

  for (const field of GENOME_LAYOUT) {
    const mask = ((1n << BigInt(field.bits)) - 1n) << BigInt(field.offset);
    const desde = rng.bool() ? menor : mayor;
    hijo |= desde & mask;
    herencia[field.key] = desde === seedA ? "A" : "B";
  }

  // Mutación, bit a bit. La probabilidad por bit se reparte para que el
  // promedio sea MUTACIONES_ESPERADAS sobre el genoma entero.
  let mutaciones = 0;
  const probabilidad = MUTACIONES_ESPERADAS / 64;
  for (let bit = 0; bit < 64; bit++) {
    if (rng.next() >= probabilidad) continue;
    hijo ^= 1n << BigInt(bit);
    mutaciones++;

    const campo = GENOME_LAYOUT.find((f) => bit >= f.offset && bit < f.offset + f.bits);
    if (campo) herencia[campo.key] = "mutado";
  }

  return { seed: hijo, herencia, mutaciones };
}

export type Elegibilidad = { puede: true } | { puede: false; motivo: string };

/**
 * ¿Esta criatura está en condiciones de cruzar?
 *
 * Se separa de `puedenCruzar` para poder decirle al jugador cuál de las dos es
 * la que no puede, y por qué.
 */
export function elegibilidad(criatura: CreatureState, nowMs: number): Elegibilidad {
  if (criatura.etapa !== "adulto") {
    return { puede: false, motivo: "Todavía no terminó de crecer." };
  }
  if (criatura.letargico) {
    return { puede: false, motivo: "Está en letargo. Primero hay que reconectar con ella." };
  }
  if (criatura.stats.salud < CRUZA_MIN_SALUD) {
    return { puede: false, motivo: "No está lo bastante sana." };
  }
  if (criatura.stats.vinculo < CRUZA_MIN_VINCULO) {
    return { puede: false, motivo: "Todavía no hay suficiente vínculo con ella." };
  }

  const ultima = criatura.ultimaCruzaMs;
  if (ultima !== null && nowMs - ultima < CRUZA_COOLDOWN_MS) {
    const faltanMs = CRUZA_COOLDOWN_MS - (nowMs - ultima);
    const horas = Math.ceil(faltanMs / 3_600_000);
    return { puede: false, motivo: `Necesita descansar. Faltan ${horas} h.` };
  }

  return { puede: true };
}

/** ¿Se puede cruzar este par? */
export function puedenCruzar(a: CreatureState, b: CreatureState, nowMs: number): Elegibilidad {
  if (a.id === b.id) {
    return { puede: false, motivo: "Hace falta otra criatura." };
  }

  const ea = elegibilidad(a, nowMs);
  if (!ea.puede) return ea;

  const eb = elegibilidad(b, nowMs);
  if (!eb.puede) return eb;

  return { puede: true };
}

/** Marca a una criatura como recién cruzada, sin mutar la original. */
export function marcarCruzada(criatura: CreatureState, nowMs: number): CreatureState {
  return { ...criatura, stats: { ...criatura.stats }, ultimaCruzaMs: nowMs };
}

/**
 * A quién salió el hijo, en texto.
 *
 * Cuenta genes, no bits: lo que el jugador percibe es "tiene la cara del padre",
 * y eso se corresponde con campos enteros.
 */
export function describirHerencia(cruza: Cruza): string {
  const valores = Object.values(cruza.herencia);
  const deA = valores.filter((v) => v === "A").length;
  const deB = valores.filter((v) => v === "B").length;

  if (cruza.mutaciones >= 5) return "Salió con bastante de suyo. Mutó mucho.";
  if (deA > deB * 2) return "Salió casi calcado al primero.";
  if (deB > deA * 2) return "Salió casi calcado al segundo.";
  return "Salió una mezcla pareja de los dos.";
}

File: gdext/src/breeding.cpp
#include "breeding.h"

#include "rng.h"

#include <cmath>

namespace petbits {

std::string_view nombreOrigen(Origen o) {
    switch (o) {
        case Origen::A:      return "A";
        case Origen::B:      return "B";
        case Origen::Mutado: return "mutado";
    }
    return "?";
}

const std::array<CampoGenoma, 14>& camposGenoma() {
    // Los catorce campos cubren los 64 bits sin huecos ni solapes, así que el
    // bucle de mutación siempre encuentra a quién culpar por un bit dado vuelta.
    static const std::array<CampoGenoma, 14> campos = {{
        {"lineage", 0, 4},     {"bodyShape", 4, 4},    {"eyes", 8, 4},
        {"mouth", 12, 4},      {"appendages", 16, 4},  {"pattern", 20, 4},
        {"hue", 24, 8},        {"paletteMode", 32, 3}, {"temperament", 35, 3},
        {"metabolism", 38, 3}, {"affinity", 41, 3},    {"proportion", 44, 4},
        {"statBias", 48, 8},   {"mutation", 56, 8},
    }};
    return campos;
}

Cruza cruzar(Seed seedA, Seed seedB, int64_t nonce) {
    // Se ordenan para sembrar el azar, y eso es lo que la vuelve simétrica:
    // cruzar A con B da lo mismo que cruzar B con A.
    const Seed menor = (seedA <= seedB) ? seedA : seedB;
    const Seed mayor = (seedA <= seedB) ? seedB : seedA;

    // El seed va en DECIMAL. En el TS la etiqueta se arma interpolando el bigint,
    // y esa interpolación rinde base diez. En hexadecimal el hash daría otro
    // número y el hijo saldría distinto, sin que nada más se rompa.
    const std::string etiqueta =
        "cruza:" + seedADecimal(mayor) + ":" + std::to_string(nonce);
    Rng rng(deriveSeed(menor, etiqueta));

    Cruza r;
    const std::array<CampoGenoma, 14>& campos = camposGenoma();

    for (size_t i = 0; i < campos.size(); ++i) {
        const CampoGenoma& campo = campos[i];
        const Seed mascara = ((1ULL << campo.bits) - 1ULL) << campo.offset;

        const Seed desde = rng.boolCon() ? menor : mayor;
        r.seed |= desde & mascara;

        // Contra seedA, no contra el menor: si seedA es el mayor de los dos, la
        // etiqueta "A" sale justamente cuando desde es el mayor. Y con dos padres
        // de genoma idéntico, todo sale "A". Es lo que hace el TS.
        r.herencia[i] = (desde == seedA) ? Origen::A : Origen::B;
    }

    // Mutación, bit a bit. La probabilidad por bit se reparte para que el
    // promedio sea MUTACIONES_ESPERADAS sobre el genoma entero.
    //
    // Los 64 bits se recorren SIEMPRE. Cada vuelta consume un número del PRNG,
    // así que cortar antes correría la secuencia y cambiaría los hijos.
    const double probabilidad = MUTACIONES_ESPERADAS / 64.0;
    for (int bit = 0; bit < 64; ++bit) {
        if (rng.next() >= probabilidad) {
            continue;
        }
        r.seed ^= (1ULL << bit);
        ++r.mutaciones;

        for (size_t i = 0; i < campos.size(); ++i) {
            if (bit >= campos[i].offset && bit < campos[i].offset + campos[i].bits) {
                // El primero que lo contiene, igual que el find() del TS.
                r.herencia[i] = Origen::Mutado;
                break;
            }
        }
    }

    return r;
}

Elegibilidad elegibilidad(const CreatureState& criatura, int64_t nowMs) {
    // El orden es el del TS y no da lo mismo: solo se informa el primer motivo,
    // y "está en letargo" es más útil que "no está sana" cuando pasan las dos.
    if (criatura.etapa != Stage::Adulto) {
        return {false, "Todavía no terminó de crecer."};
    }
    if (criatura.letargico) {
        return {false, "Está en letargo. Primero hay que reconectar con ella."};
    }
    if (criatura.stats.salud < CRUZA_MIN_SALUD) {
        return {false, "No está lo bastante sana."};
    }
    if (criatura.stats.vinculo < CRUZA_MIN_VINCULO) {
        return {false, "Todavía no hay suficiente vínculo con ella."};
    }

    if (criatura.ultimaCruzaMs.has_value()) {
        const int64_t desde = nowMs - *criatura.ultimaCruzaMs;
        if (desde < CRUZA_COOLDOWN_MS) {
            const int64_t faltanMs = CRUZA_COOLDOWN_MS - desde;
            // Hacia arriba y en punto flotante, como el Math.ceil del TS. Con
            // división entera, "faltan 0 h" aparecería durante la última hora.
            const int64_t horas =
                static_cast<int64_t>(std::ceil(static_cast<double>(faltanMs) / 3600000.0));
            return {false, "Necesita descansar. Faltan " + std::to_string(horas) + " h."};
        }
    }

    return {true, ""};
}

Elegibilidad puedenCruzar(const CreatureState& a, const CreatureState& b, int64_t nowMs) {
    if (a.id == b.id) {
        return {false, "Hace falta otra criatura."};
    }

    const Elegibilidad ea = elegibilidad(a, nowMs);
    if (!ea.puede) {
        return ea;
    }

    const Elegibilidad eb = elegibilidad(b, nowMs);
    if (!eb.puede) {
        return eb;
    }

    return {true, ""};
}

CreatureState marcarCruzada(const CreatureState& criatura, int64_t nowMs) {
    CreatureState copia = criatura;
    copia.ultimaCruzaMs = nowMs;
    return copia;
}

std::string describirHerencia(const Cruza& cruza) {
    int deA = 0;
    int deB = 0;
    for (Origen o : cruza.herencia) {
        if (o == Origen::A) ++deA;
        if (o == Origen::B) ++deB;
    }

    // Los mutados no cuentan para ninguno de los dos lados, así que con muchas
    // mutaciones deA y deB pueden sumar bastante menos de catorce. Por eso el
    // caso de "mutó mucho" se pregunta primero.
    if (cruza.mutaciones >= 5) return "Salió con bastante de suyo. Mutó mucho.";
    if (deA > deB * 2) return "Salió casi calcado al primero.";
    if (deB > deA * 2) return "Salió casi calcado al segundo.";
    return "Salió una mezcla pareja de los dos.";
}

} // namespace petbits
